Karumi/Shot
# Execute tests and pull screenshots

### :tophat: What is the goal?

Continue shot development adding a new feature to be able to pull the screenshot from the Android device connected to our computer.

**If your review this PR you'll see how the paradigm used is OOP and not FP**. This is because I'd like to use [this library](http://frees.io/) named ``Freestyle`` but I couldn't get Scala macros working on Gradle.

### How is it being implemented?

The steps followed to implement this feature are:

* Add a Scala code formatter and integrate it into our TravisCI configuration. You can find more information about scala fmt [here](http://scalameta.org/scalafmt/).
* Create two main abstractions we will use for executing adb commands and interact with the view interface by printing a message in the standard output. The abstractions names are ``Adb`` and ``Console``.
* Connect our ``ShotPlugin`` and different tasks with a new domain model named ``Shot`` using ``Adb`` and ``Console`` as collaborators.
* Implement all the logic needed to clean and pull screenshots from the connected Android device.
* Create a new extension needed to get the application package easily. This could be improved in the future if needed.
* Add some libraries needed to mock dependencies in our unitary tests

### How can it be tested?

Clone this branch and execute these commands:

```
./gradlew install
cd shot-consumer
./gradlew executeScreenshotTests
```

You should see a new folder inside ``shot-consumer/app/screenshots/screenshots-default`` with a bunch of ``png`` files inside and some metadata files.

### Next tasks?

* Refactor part of the system initialization to use [cake pattern](http://www.cakesolutions.net/teamblogs/2011/12/19/cake-pattern-in-depth).
* Start working on the error handling to be able to break a task execution when needed by returning an error modeled using ``Either`` and sealed traits.
* Start comparing images and getting the difference between every pair of screenshots. To be able to do this we will have to parse the metadata file we've pulled from the connected device.
